File: Python/Tema9_Ejercicio6.py
"""
TEMA 9: PROGRAMACIÓN ORIENTADA A OBJETOS
EJERCICIO 6: ENCAPSULACIÓN
===========================

Controla el acceso a atributos privados
"""

# ATRIBUTOS PRIVADOS
print("=== PRIVADOS ===\n")

class Empleado:
    def __init__(self, nombre, salario):
        self.nombre = nombre
        self.__salario = salario
    
    def obtener_salario(self):
        return self.__salario
    
    def aumentar_salario(self, porcentaje):
        if porcentaje > 50:
            print("Error: aumento máximo 50%")
            return
        self.__salario *= (1 + porcentaje / 100)

emp = Empleado("Carlos", 2000)
print(f"Salario: ${emp.obtener_salario()}")

emp.aumentar_salario(10)
print(f"Con 10% aumento: ${emp.obtener_salario():.2f}")

emp.aumentar_salario(60)

# MÉTODOS DE VALIDACIÓN
print("\n=== VALIDACIÓN ===\n")

class BancoCuenta:
    def __init__(self, numero, saldo=0):
        self.__numero = numero
        self.__saldo = saldo
        self.__historial = []
    
    def depositar(self, cantidad):
        if cantidad <= 0:
            print("Error: cantidad debe ser positiva")
            return
        self.__saldo += cantidad
        self.__historial.append(f"Depósito: ${cantidad}")
    
    def retirar(self, cantidad):
        if cantidad <= 0:
            print("Error: cantidad debe ser positiva")
            return
        if cantidad > self.__saldo:
            print("Error: fondos insuficientes")
            return
        self.__saldo -= cantidad
        self.__historial.append(f"Retiro: ${cantidad}")
    
    def obtener_saldo(self):
        return self.__saldo

cuenta = BancoCuenta("123456", 1000)
print(f"Saldo inicial: ${cuenta.obtener_saldo()}")

cuenta.depositar(500)
print(f"Después depositar: ${cuenta.obtener_saldo()}")

cuenta.retirar(200)
print(f"Después retirar: ${cuenta.obtener_saldo()}")

cuenta.retirar(2000)

"""
SALIDA ESPERADA:
=== PRIVADOS ===

Salario: $2000
Con 10% aumento: $2200.00
Error: aumento máximo 50%

=== VALIDACIÓN ===

Saldo inicial: $1000
Después depositar: $1500
Después retirar: $1300
Error: fondos insuficientes
"""

File: JS/Tema9_Ejercicio6.js
/*
 * TEMA 9: PROGRAMACIÓN ORIENTADA A OBJETOS - EJERCICIO 6
 * ========================================================
 * Encapsulación: control de acceso.
 * 
 * EXPLICACIÓN:
 * - Ocultar detalles internos
 * - Exposición controlada a través de métodos
 * - Mantener invariantes de la clase
 */

class Empleado {
    constructor(nombre, salarioBase) {
        this.nombre = nombre;
        this.salarioBase = salarioBase;
        this.bonificacion = 0;
    }
    
    agregarBonificacion(cantidad) {
        if (cantidad >= 0 && cantidad <= this.salarioBase * 0.5) {
            this.bonificacion = cantidad;
            return true;
        }
        console.warn(`Bonificación debe estar entre 0 y ${this.salarioBase * 0.5}`);
        return false;
    }
    
    obtenerSalarioTotal() {
        return this.salarioBase + this.bonificacion;
    }
    
    obtenerDetalle() {
        return `${this.nombre}: Salario base $${this.salarioBase} + Bonificación $${this.bonificacion} = $${this.obtenerSalarioTotal()}`;
    }
}

class Producto {
    constructor(nombre, precioBase) {
        this.nombre = nombre;
        this.precioBase = precioBase;
        this.descuento = 0;
    }
    
    aplicarDescuento(porcentaje) {
        if (porcentaje >= 0 && porcentaje <= 100) {
            this.descuento = porcentaje;
            return true;
        }
        console.warn("Descuento debe estar entre 0 y 100");
        return false;
    }
    
    obtenerPrecioFinal() {
        return this.precioBase * (1 - this.descuento / 100);
    }
    
    obtenerInfo() {
        const precioFinal = this.obtenerPrecioFinal();
        const ahorrado = this.precioBase - precioFinal;
        return `${this.nombre}: $${this.precioBase} - ${this.descuento}% = $${precioFinal.toFixed(2)} (Ahorro: $${ahorrado.toFixed(2)})`;
    }
}

console.log("=== ENCAPSULACIÓN ===\n");

const emp1 = new Empleado("Carlos", 2000);
console.log("Empleados:");
console.log(emp1.obtenerDetalle());

emp1.agregarBonificacion(300);
console.log(emp1.obtenerDetalle());

const prod = new Producto("Laptop", 1000);
console.log("\nProductos:");
console.log(prod.obtenerInfo());

prod.aplicarDescuento(20);
console.log(prod.obtenerInfo());

/*
 * SALIDA ESPERADA:
 * === ENCAPSULACIÓN ===
 * 
 * Empleados:
 * Carlos: Salario base $2000 + Bonificación $0 = $2000
 * Carlos: Salario base $2000 + Bonificación $300 = $2300
 * 
 * Productos:
 * Laptop: $1000 - 0% = $1000.00 (Ahorro: $0.00)
 * Laptop: $1000 - 20% = $800.00 (Ahorro: $200.00)
 */
